walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,67 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.ImmutableList; 
 
+/** Finds list of all potential calendar time slots to schedule a meeting with the given 
+ * MeetingRequest constraints. Accounts for optional attendees when possible. */
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ImmutableList<TimeRange> conflicting = getRelevantEvents(request.getAttendees(), events);
+    ImmutableList<TimeRange> times = getMeetingTimes(request.getDuration(), conflicting);
+    
+    Iterable<String> allAttendeesIter = 
+        Iterables.unmodifiableIterable(Iterables.concat(request.getAttendees(),
+                                                        request.getOptionalAttendees()));
+    Collection<String> allAttendees = Lists.newArrayList(allAttendeesIter);
```

Comment:
Let's try to stick to using as few different container types as we need to. Iterables and Collections in this case do the same thing as ImmutableList, so let's try to convert these into ImmutableLists. Similarly, let's have your methods below take ImmutableLists as parameters. Or maybe, ImmutableSet would be slightly better for efficiency for objects you don't want duplicates of (like attendees maybe).

The ImmutableList.copyOf() method is an easy way to turn any collection into an ImmutableList. https://guava.dev/releases/23.0/api/docs/com/google/common/collect/ImmutableList.html#copyOf-java.util.Collection-